pkg/interop/native/ledger/transaction_signer.go

```diff
@@ -0,0 +1,145 @@
+package ledger
+
+import "github.com/nspcc-dev/neo-go/pkg/interop"
+
+// TransactionSigner represent the signer of a NEO transaction. It's similar to
+// Signer class in Neo .net framework.
+type TransactionSigner struct {
+	// Bytes represents the binary serialized form of the given signer.
+	Bytes []byte
+	// Account represents the account (160 bit BE value in a 20 byte slice) of
+	// the given signer.
+	Account interop.Hash160
+	// Scopes represents a set of witness flags for the given signer.
+	Scopes SignerScope
+	// Contracts represents the set of contract hashes (160 bit BE value in a 20
+	// byte slice) allowed to be called by the signer.
+	AllowedContracts []interop.Hash160
+	// AllowedGroups represents the set of contract groups (ecdsa public key
+	// bytes in a 32 byte slice) allowed to be called by the signer.
+	AllowedGroups []interop.PublicKey
+	// Rules represents a rule-based witness scope of the given signer.
+	Rules []WitnessRule
+}
+
+// SignerScope represents a signer's witness scope.
+type SignerScope byte
+
+// Various witness scopes.
+const (
+	// None specifies that no contract was witnessed. Only signs the transaction.
+	None SignerScope = 0
+	// CalledByEntry means that the witness is valid only when EntryScriptHash
+	// equals to the CallingScriptHash.
+	CalledByEntry SignerScope = 0x01
+	// CustomContracts define custom hash for contract-specific witness.
+	CustomContracts SignerScope = 0x10
+	// CustomGroups define custom public key for group members.
+	CustomGroups SignerScope = 0x20
+	// Rules is a set of conditions with boolean operators.
+	Rules SignerScope = 0x40
+	// Global allows this witness in all contexts. This cannot be combined with
+	// other flags.
+	Global SignerScope = 0x80
+)
+
+// WitnessRule represents a single rule for Rules witness scope.
+type WitnessRule struct {
+	Action WitnessAction
+	// Condition can be cast either to WitnessConditionWithType or to
+	// WitnessConditionWithTypeAndValue.
+	Condition interface{}
```

Comment:
That's the case, I also like this solution, but are you sure that we can always use a structure with two fields (type and value) as Condition? I thought that such script will be FAULTed in case of CalledByEntry, because CalledByEntry condition contains single stackitem in the Condition array, and we'll try to cast it to a struct with two fields (two stackitems are expected). So in case of CalledByEntry the array is not stackitem+null, it's just a single stackitem.

Reply:
You can make a test for it, I think it'll work.

Reply:
I've implemented a test for that, see `TestLedger_GetTransactionSignersInteropAPI`. In the test we deploy a contract. The contract calls GetSigners method of native Ledger and checks that returned Signers contain all fields.

Turns out that this test is failing on both implementations (with `Condition interface{}` and `Condition WitnessConditionWithTypeAndValue`.

Reply:
I found a bug :D

Reply:
> casting to some type != accessing the second field

You were right, implemented. And the test also passes.